### ISSUE TITLE
Hypre matrix and vector operation

### ISSUE THREAD

**Issue by martemyev** [mfem/mfem#67]
Hi All,

I've got a question about parallel matrix-vector operations, and since MFEM is tightly connected to Hypre I thought I may find an answer here. The problem is quite simple - I've got several matrices in CSR format that I build myself, and I do it sequentially (ideally I'd do it in parallel, but it's not that important right now). My goal is then distribute (scatter) these matrices along with some vectors (like RHS) among available processes via MPI, and then do linear algebra operations (mostly mat-vec multiplication and vector addition, subtraction) in parallel. I understand, that this approach is only partially parallel as the building of the matrices is sequential, but that's fine.

So, the question is: Can I initialize parallel Hypre matrices and vectors from my matrices and vectors (i.e. distribute them), and then do some linear algebra operations in parallel?

Thanks,
Mikhail

**Comment by tzanio** [comment]
Hi Mikhail,

Will a block-diagonal hypre matrix with blocks given by your CSR matrices work for your needs?

Tzanio

**Comment by martemyev** [comment]
Hi Tzanio,
Thanks for the suggestion, but I think it's not what I'd need. My goal is to cut a global sparse matrix in pieces and send these pieces to different MPI processes, so that mat-vec operations would be performed in parallel. For example, we can split the matrix in terms of rows, i.e. 0:N-1 rows would go to process 0, then N:2*N-1 rows would go to process 1, and so on. That's what I meant by distributing a matrix among processes. I guess that's not difficult to implement, but I thought there might be already a solution within Hypre.

Thanks,
Mikhail

**Comment by trb413** [comment]
Hi Mikhail,

If you have not built hypre with the HYPRE_NO_GLOBAL_PARTITION option (unfortunately, you probably have this enabled since it's the default), there is a function called "HYPRE_CSRMatrixToParCSRMatrix()" that takes a communicator, a CSR matrix on rank 0, and optionally row/column partitionings (if they are null, Hypre creates them), and distributes this in a HYPRE_ParCSRMatrix. There is a similar function for vectors: "HYPRE_VectorToParVector()". These can be found in the "parcsr_mv" directory of the hypre source. Hypre, of course, will have the kernels for mat-vecs and vector operations that you mentioned

There's also an mfem::HypreParMatrix constructor that wraps this function (nothing for vectors, unfortunately, but it's substantially easier to do vectors), but it, too, is bound by this "HYPRE_NO_GLOBAL_PARTITION" restriction.

If you have a globally replicated matrix/vector, you can do this even more easily with no communication -- basically each rank in the communicator forgets the rows of the matrix it doesn't own. I have some code that does this. If you'd like, I can clean up the code and post it here.

Cheers,
Tom

**Comment by martemyev** [comment]
Hi Tom,

That seems to be exactly what I need. Although I think I got the idea, an example is always most illustrative, so it would be great if you'd post it here.

Thanks,
Mikhail

**Comment by trb413** [comment]
Hi Mikhail,

You might try something like this.

My disclaimers are these: I pulled pieces of this together from a bigger chunk of C++11 code and tried to translate it to mostly-C-friendly code (completely C-friendly if C11), so it might not be the cleanest (and I hope there are no indexing errors...); I am not a hypre developer, so this might not be the "endorsed" way to do things; this code does no internal error checking (Hypre functions return an error code, which I never bother to check).

Please let me know if you have any questions or problems or find any bugs.

Cheers,
Tom

```cxx
#ifndef DISTRIBUTEGLOBALLYREPLICATEDHYPRE_HPP_
#define DISTRIBUTEGLOBALLYREPLICATEDHYPRE_HPP_

#include <HYPRE.h>
#include <_hypre_parcsr_mv.h>

// Inputs:
//   comm = MPI Communicator you want to distribute over
//   serial_I = "row_ptr" pointer in CSR structure
//   serial_J = "col_ind" pointer in CSR structure
//   serial_Data = "values" pointer in CSR structure
//   my_row_starts = array of length 3:
//      [my_first_row,one_past_my_last_row,global_num_rows]
//   my_col_starts = array of length 3:
//      [my_first_col,one_past_my_last_col,global_num_cols]
//
//   where,
//
//   global_num_rows = number of rows in global matrix
//   global_num_cols = number of columns in global matrix
//
//   Assumes caller has ownership of my_row_starts and
//   my_col_starts. Further assumes that caller is using
//   HYPRE_NO_GLOBAL_PARTITION.
//
// Returns a newly created hypre_ParCSRMatrix. Caller is responsible
// for destroying the object.
hypre_ParCSRMatrix* DistributeGloballyReplicatedMatrix(
    MPI_Comm comm,
    HYPRE_Int* serial_I, HYPRE_Int* serial_J, HYPRE_Complex* serial_Data,
    HYPRE_Int* my_row_starts, HYPRE_Int* my_col_starts)
{

    const HYPRE_Int global_num_rows = my_row_starts[2],
        global_num_cols = my_col_starts[2];
    
    const HYPRE_Int my_first_row = my_row_starts[0],
        my_last_row = my_row_starts[1];

    const HYPRE_Int my_first_col = my_col_starts[0],
        my_last_col = my_col_starts[1];
    
    const HYPRE_Int my_num_rows = my_last_row - my_first_row;

    // Create the offd_col_map first
    HYPRE_Int * global_idx_marker =
        (HYPRE_Int *) malloc(global_num_cols*sizeof(HYPRE_Int));

    HYPRE_Int num_shared_cols = 0, diag_nnz = 0, offd_nnz = 0;
    for (HYPRE_Int index = serial_I[my_first_row];
         index < serial_I[my_last_row];
         ++index)
    {
        const HYPRE_Int& the_global_column_index = serial_J[index];

        if ( (the_global_column_index < my_first_col) ||
             (the_global_column_index >= my_last_col) )
        {
            // The index is outside my owned range; mark and up count
            if (!global_idx_marker[the_global_column_index])
                global_idx_marker[the_global_column_index] = ++num_shared_cols;

            // Up nnz count
            offd_nnz++;
        }
        else
            diag_nnz++;
    }
        
    // Create the new matrix
    hypre_ParCSRMatrix * parallel_matrix = hypre_ParCSRMatrixCreate(
        comm,global_num_rows,global_num_cols,
        my_row_starts,my_col_starts,
        num_shared_cols,diag_nnz,offd_nnz);

    // In my case, I keep the partitioning arrays
    hypre_ParCSRMatrixOwnsRowStarts(parallel_matrix) = 0;
    hypre_ParCSRMatrixOwnsColStarts(parallel_matrix) = 0;
    
    // Initialize the matrix -- allocates Diag and Offd
    hypre_ParCSRMatrixInitialize(parallel_matrix);

    // Fill col_map_offd
    HYPRE_Int * col_map_offd = hypre_ParCSRMatrixColMapOffd(parallel_matrix);

    num_shared_cols = 0;
    for (size_t global_idx = 0;
         global_idx < global_num_cols;
         ++global_idx)
    {
        if (global_idx_marker[global_idx] > 0)
        {
            col_map_offd[num_shared_cols] = global_idx;
            global_idx_marker[global_idx] = ++num_shared_cols;
            // Note: this will mark each shared global idx with a
            // positive number, so
            //
            // local_idx = global_idx_marker[global_idx] - 1.
        }
    }
    
    // Copy the matrix data
    hypre_CSRMatrix * diag = hypre_ParCSRMatrixDiag(parallel_matrix),
        * offd = hypre_ParCSRMatrixOffd(parallel_matrix);

    HYPRE_Int * diag_i = hypre_CSRMatrixI(diag),
        * diag_j = hypre_CSRMatrixJ(diag);
    HYPRE_Complex * diag_data = hypre_CSRMatrixData(diag);
    
    HYPRE_Int * offd_i = hypre_CSRMatrixI(offd),
        * offd_j = hypre_CSRMatrixJ(offd);
    HYPRE_Complex * offd_data = hypre_CSRMatrixData(offd);

    diag_nnz = 0; offd_nnz = 0;
    for (HYPRE_Int global_row = my_first_row, local_row = 0;
         global_row < my_last_row;++global_row,++local_row)
    {
        diag_i[local_row] = diag_nnz;
        offd_i[local_row] = offd_nnz;
        
        for (HYPRE_Int j_idx = serial_I[global_row];
             j_idx < serial_I[global_row+1]; ++j_idx)
        {
            const HYPRE_Int& global_idx = serial_J[j_idx];

            if ((global_idx < my_first_col) || (global_idx >= my_last_col))
            {
                // Column is shared, add to offd
                offd_j[offd_nnz] = global_idx_marker[global_idx] - 1;
                offd_data[offd_nnz++] = serial_Data[j_idx];
            }
            else
            {
                // Column is owned, add to diag
                diag_j[diag_nnz] = global_idx - my_first_col;
                diag_data[diag_nnz++] = serial_Data[j_idx];
            }
        }
    }
    diag_i[my_num_rows] = diag_nnz;
    offd_i[my_num_rows] = offd_nnz;

    // Hypre wants the diagonal entry to come first.
    HYPRE_Complex tmp_data;
    for (HYPRE_Int local_row=0; local_row < my_num_rows; ++local_row)
    {
        HYPRE_Int first_col = diag_i[local_row];
        for (HYPRE_Int j_idx = first_col; j_idx < diag_i[local_row+1]; ++j_idx)
        {
            if (diag_j[j_idx] == local_row)
            {
                // Swap the column indices
                diag_j[j_idx] = diag_j[first_col];
                diag_j[first_col] = local_row;

                // Swap the data
                tmp_data = diag_data[j_idx];
                diag_data[j_idx] = diag_data[first_col];
                diag_data[first_col] = tmp_data;
                break;
            }
        }
    }

    // Now the hypre_ParCSRMatrix should be built. Create the commpkg:
    hypre_MatvecCommPkgCreate(parallel_matrix);

    free(global_idx_marker);
    
    return parallel_matrix;
}// DistributeGloballyReplicatedMatrix


// Inputs:
//   comm = MPI Communicator over which you want to distribute
//   global_num_entries = length of global vector
//   serial_x = vector data
//   partitioning = array of length 2:
//      [my_first_entry,one_past_my_last_entry]
//
//   Assumes caller has ownership of partitioning. Further assumes
//   that caller is using HYPRE_NO_GLOBAL_PARTITION.
//
// Returns a newly created hypre_ParVector. Caller is responsible
// for destroying the object.
hypre_ParVector DistributeGloballyReplicatedVector(
    MPI_Comm comm, HYPRE_Int global_num_entries, HYPRE_Complex* serial_x,
    HYPRE_Int* partitioning)
{
    hypre_ParVector * parallel_x = hypre_ParVectorCreate(
        comm,global_num_entries,partitioning);

    HYPRE_Int my_first_entry = partitioning[0],
        my_last_entry = partitioning[1];
    
    hypre_ParVectorOwnsPartitioning(parallel_x) = 0;

    hypre_ParVectorInitialize(parallel_x);

    hypre_Vector * local_x = hypre_ParVectorLocalVector(parallel_x);
    HYPRE_Complex * x_data = hypre_VectorData(local_x);

    // Copy the data
    for (HYPRE_Int idx = my_first_entry; idx < my_last_entry; ++idx)
        x_data[idx - my_first_entry] = serial_x[idx];

    return parallel_x;
}

#endif /* DISTRIBUTEGLOBALLYREPLICATEDHYPRE_HPP_ */
```

**Comment by tzanio** [comment]
@trb413, thanks for the detailed example!
@martemyev, can we close this issue?

**Comment by martemyev** [comment]
Sure. Thanks for the example!

**Comment by martemyev** [comment]
Hi All,

I was trying to combine two matrices with different parallel distribution and failed. It's somehow related to what I was asking in this thread earlier, so I decided to reopen it.
So, here is what I'm doing:
1. I build a parallel HypreParMatrix by means of MFEM.
2. I compute some small matrices R_i for my task in parallel
3. I need to build a global R matrix, and do the multiplication R^T A R
The thing is that I want the global R matrix also be parallel, and I make it so (hopefully correctly) by having each process to have a stripe of the global R matrix. Each stripe consists of several rows of that global matrix, and in each stripe there are only R_i matrix computed by that process, i.e. I don't need to send-receive R_i matrices. The R_i matrices are not blocks, however, so I can't put everything in a block-diagonal parallel sparse matrix.

The code below shows the idea of what I'm doing. For simplicity I assume R_i are rectangular matrices filled with ones.

```cxx
#include "config.hpp"
#include "mfem.hpp"

using namespace mfem;
using namespace std;

int main(int argc, char **argv) {
  MPI_Init(&argc, &argv);

  int myid, nproc;
  MPI_Comm_rank(MPI_COMM_WORLD, &myid);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);

  const int nx = 100;
  const int ny = 100;
  const bool generate_edges = 1;
  const int order = 1;
  Mesh mesh(nx, ny, Element::QUADRILATERAL, generate_edges);
  const int dim = mesh.Dimension();
  ParMesh par_mesh(MPI_COMM_WORLD, mesh);

  H1_FECollection fec(order, dim);
  ParFiniteElementSpace fespace(&par_mesh, &fec);

  ParBilinearForm a(&fespace);
  a.AddDomainIntegrator(new DiffusionIntegrator());
  a.Assemble();
  a.Finalize();
  HypreParMatrix *A = a.ParallelAssemble();

  const int ndofs = fespace.GlobalTrueVSize();
  if (myid == 0)
    cout << "ndofs " << ndofs << "\n";

  srand(time(0));

  // here comes some matrix R - filled with 1 for this example
  const int R_nrows = 100;
  const int R_ncols = ndofs;
  // distribute the matrix among processes by stripes of several rows
  const int min_nrows  = R_nrows / nproc;
  const int extra_rows = R_nrows % nproc;
  const int start_row  = min_nrows * myid + (extra_rows < myid ? extra_rows : myid);
  const int end_row    = start_row + min_nrows + (extra_rows > myid);
  const int my_nrows   = end_row - start_row;
  cout << "myid " << myid << ": start_row " << start_row << " end_row "
       << end_row << endl;

  int *Ri = new int[my_nrows + 1];
  int *Rj = new int[my_nrows * R_ncols];
  double *Rdata = new double[my_nrows * R_ncols];

  Ri[0] = 0;
  for (int i = 1; i < my_nrows + 1; ++i)
    Ri[i] = Ri[i - 1] + R_ncols;

  for (int i = 0; i < my_nrows; ++i) {
    for (int j = 0; j < R_ncols; ++j) {
      Rj[i*R_ncols + j] = j;
      Rdata[i*R_ncols + j] = 1.0;
    }
  }

  MFEM_VERIFY(HYPRE_AssumedPartitionCheck(), "HYPRE_NO_GLOBAL_PARTITION must be ON");
  int myRrows[] = { start_row, end_row };
  int myRcols[] = { 0, R_ncols };

  HypreParMatrix R_global(MPI_COMM_WORLD, my_nrows, R_nrows, R_ncols,
                          Ri, Rj, Rdata, myRrows, myRcols);
  HypreParMatrix *R_global_T = R_global.Transpose();

  HypreParMatrix *K = RAP(A, R_global_T); // K = R^T A R

  delete K;
  delete R_global_T;
  delete[] Rdata;
  delete[] Rj;
  delete[] Ri;
  delete A;

  MPI_Finalize();

  return 0;
}
```

With only one process that works fine, but with more it doesn't. I suspect there is an error in compatibility of number of columns of matrix R^T and rows of A - I mean not global ones, but the ones possessed by different processes.

Any ideas about how to deal with this case is highly appreciated!

Thanks,
Mikhail

**Comment by trb413** [comment]
Hi Mikhail,

The issue here is that you don't partition your columns. You appear to be
passing the array {0,ndofs} as the "cols" pointer to the HypreParMatrix
constructor, which essentially tells hypre that each process owns all of
the columns. So I would suggest looking at partitioning the columns in the
same way that you've partitioned the rows. You have at least two good
options here:

1. If you have a general matrix that you're building out of thin air, you
have to create these yourself. You can use the same algorithm you used to
create start_row and end_row. If you don't mind calling hypre functions
directly, you can call "hypre_GenerateLocalPartitioning()" to generate this
partitioning for you. I doubt that this is "officially sanctioned" since it
starts with a lower-case "hypre", but this handy helper function basically
does exactly what you do to create start_row and end_row.

2. In your case, you have a distributed matrix lying around that you can
borrow from. In this particular case, I suggest that you just "borrow" the
(column) partition from "A", which you can get by calling "A->ColPart()"
(column partition because you want the columns of A to be in the range of
R, it looks like). This (a) is easiest, (b) requires the least thought, and
(c) is least likely to cause problems when you call that RAP function.
Code-wise, replace your "R_global" initialization line with:

HypreParMatrix R_global(MPI_COMM_WORLD, my_nrows, R_nrows, R_ncols,
                          Ri, Rj, Rdata, myRrows, A->ColPart());

and your code will run just fine (as to whether or not it does the right
thing, you'll have to tell me...).

A short explanation: The constructor you are using is for converting a
matrix in a parallel CSR format (such as in various Trilinos packages or
some sparse direct solvers, e.g. out of LBNL) to the diag/offd dual-CSR
format used by Hypre (and petsc). In Hypre, each process owns contiguous
sets of rows, as you note in your description, but also contiguous sets of
columns. The process stores its rows in two pieces: the chunk of columns it
owns (the CSR matrix called "diag") and the chunk of columns it "shares"
(the CSR matrix called "offd", or what I call "everything else"). Thus the
user is expected to provide both the range of owned rows (usually called
"row_starts", but seemingly just called "rows" in this constructor) and the
range of owned columns (usually called "col_starts", but just called "cols"
in this constructor). These ranges should be given in global IDs.

Anyway, I hope this helps. Let me know if you have any other questions.

Cheers,
Tom

**Comment by martemyev** [comment]
Hi Tom,

Thanks a lot for the explanation! Now many things became clearer. The second approach you mentioned indeed runs fine, but I don't really know if that's correct until I compare the serial and parallel matrices and the result of the RAP multiplication. Ideally, I'd use PrintMatlab function to examine the matrices in Matlab, but I doubt it works for HypreParMatrix. Is there any approach that I could use? For testing purposes, something like building a serial sparse matrix on process 0 from the HypreParMatrix distributed over several processes, and then using PrintMatlab would be great. But I don't see anything like that in the MFEM.

Thanks,
Mikhail

**Comment by trb413** [comment]
Hi Mikhail,

There are two approaches to this.

First, you can create a file name, say `std::string filename = "output_file";` and then call

```cxx
mfem::HypreParMatrix A;
/* stuff happens */
A->Print(filename.c_str());
```

and this will print the local pieces of the matrix to files like "output_file.xxxxx", where "xxxxx" becomes the MPI rank that wrote the file.

Alternatively, you can use a hypre function to create a hypre_CSRMatrix on every process, wrap that as a SparseMatrix, and then call PrintMatlab from one process.

DISCLAIMER: the function is "hidden" in the depths of Hypre and prefixed by "hypre_" so its interface is not guaranteed across versions. But, it's what I know, so it's what I can pass on.

```cxx
mfem::HypreParMatrix A;
/* Stuff happens */

hypre_ParCSRMatrix* A_serial = hypre_ParCSRMatrixToCSRMatrixAll(A);
// This call works because HypreParMatrix implicitly converts to hypre_ParCSRMatrix*

// This "views" the hypre_CSRMatrix as an mfem::SparseMatrix
mfem::SparseMatrix A_sparse(
    hypre_CSRMatrixI(A_serial), hypre_CSRMatrixJ(A_serial), hypre_CSRMatrixData(A_serial),
    hypre_CSRMatrixNumRows(A_serial), hypre_CSRMatrixNumCols(A_serial),
    false, false, true);

// Write to file from root process
if (myid == 0)
{
    std::ofstream outfile("output_file.mat");
    A_sparse.PrintMatlab(outfile);
    outfile.close();
}

// Cleanup, since the hypre call creates a new matrix on each process
hypre_CSRMatrixDestroy(A_serial);
```

Note that this approach will copy the entire matrix onto every process, which will kill a lot of memory. Thus, it is fine for debugging but might not be so good in real code.

Hope that helps. Let me know if you have any other questions!

Cheers,
Tom

**Comment by martemyev** [comment]
Hi Tom,

Thank you again. That's very helpful! However, when I tried to follow my typical workflow of looking at the matrices from MFEM in Matlab, I found out that something is wrong in my parallel program. Here is the code:
```cxx
#include "mfem.hpp"

using namespace mfem;
using namespace std;

void print_par_matrix_matlab(HypreParMatrix &A, const string &filename, int myid)
{
  hypre_CSRMatrix* A_serial = hypre_ParCSRMatrixToCSRMatrixAll(A);

  mfem::SparseMatrix A_sparse(
    hypre_CSRMatrixI(A_serial), hypre_CSRMatrixJ(A_serial), hypre_CSRMatrixData(A_serial),
    hypre_CSRMatrixNumRows(A_serial), hypre_CSRMatrixNumCols(A_serial),
    false, false, true);

  if (myid == 0) {
    ofstream out(filename.c_str());
    MFEM_VERIFY(out, "Cannot open file " << filename);
    A_sparse.PrintMatlab(out);
  }

  hypre_CSRMatrixDestroy(A_serial);
}

int main(int argc, char **argv) {
  MPI_Init(&argc, &argv);

  int myid, nproc;
  MPI_Comm_rank(MPI_COMM_WORLD, &myid);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);

  const int nx = 16;
  const int ny = 16;
  const bool generate_edges = 1;
  const int order = 1;
  Mesh mesh(nx, ny, Element::QUADRILATERAL, generate_edges);
  const int dim = mesh.Dimension();
  ParMesh par_mesh(MPI_COMM_WORLD, mesh);

  H1_FECollection fec(order, dim);
  ParFiniteElementSpace fespace(&par_mesh, &fec);

  ParBilinearForm a(&fespace);
  a.AddDomainIntegrator(new DiffusionIntegrator());
  a.Assemble();
  a.Finalize();
  HypreParMatrix *A = a.ParallelAssemble();

  ostringstream oss;
  oss << "a_mat_nproc" << nproc << ".dat";
  print_par_matrix_matlab(*A, oss.str(), myid);

  delete A;

  MPI_Finalize();
  return 0;
}
```
The `print_par_matrix_matlab` is basically what you suggested except the `hypre_ParCSRMatrix* A_serial` from your code was replaced by `hypre_CSRMatrix* A_serial`.
I run this code with 1, 2 and 4 processes, so I have 3 files a_mat_nprocX.dat, where X=1,2,4
Then I convert them in sparse matrices in Matlab this way:
```
load a_mat_nproc1.dat
A1 = spconvert(a_mat_nproc1);
figure; spy(A1), title('nproc 1')

load a_mat_nproc2.dat
A2 = spconvert(a_mat_nproc2);
figure; spy(A2), title('nproc 2')

load a_mat_nproc4.dat
A4 = spconvert(a_mat_nproc4);
figure; spy(A4), title('nproc 4')
```
The three plots of the sparsity pattern are attached.
I'm really confused about where the error is.
![a_matrices](https://cloud.githubusercontent.com/assets/2182432/17069134/91bc7d46-501a-11e6-9daa-dfe6a7e6f8aa.png)


Thanks,
Mikhail

**Comment by trb413** [comment]
Hi Mikhail,

Thanks for catching the "hypre_CSRMatrix" mistake! I've corrected my original comment.

I would guess that these matrices, in fact, are all "the same" in the sense that all the variables are there with the correct entries et cetera but in a permuted order. You'll note that they are all symmetric; they all have zero row-sum; they all have the same Frobenius norm. Of course these tests are not 100% guarantees, but they do suggest a reasonable possibility that the matrices you get are just different representations of the same operator (and they are implementable in about 10 seconds each...).

This appears to be caused by the choice of global IDs by MFEM, but perhaps one of the more finite-elementy MFEM people can shed some light on why this permutation occurs. @tzanio, is this just MFEM picking a different global ordering of the dofs?

If you have some reason to suspect that they are in fact not representations of the same operator, do let me know and I can dig deeper into this.

Cheers,
Tom

**Comment by martemyev** [comment]
Hi Tom,

What you're saying definitely makes sense. Let me rephrase it in the MFEM terms: in parallel finite element space the degrees of freedom have indices different from the ones that are in serial finite element space. Is that correct? Let me explain with the code:
```cxx
Mesh mesh(/*parameters*/);
FiniteElementCollection *fec = /*some collection*/;
FiniteElementSpace fespace_serial(&mesh, fec, dim);
ParMesh par_mesh(MPI_Comm, &mesh);
ParFiniteElementSpace fespace_parallel(&par_mesh, fec, dim);

Array<int> dofs_serial, dofs_parallel;
const int cellID = 10;
fespace_serial.GetElementVDofs(cellID, dofs_serial);
fespace_parallel.GetElementVDofs(cellID, dofs_parallel);

/* And here dofs_serial[i] != dofs_parallel[i] */
``` 
Do I understand it correctly?

Thanks,
Mikhail

**Comment by trb413** [comment]
Hi Mikhail,

I believe that is correct in principle. More precisely,

```cxx
mfem::Array<int> dofs_serial, dofs_parallel;
const int global_cellID = 10;
if (/* Global element 10 lives on this process */)
{
    const int local_cellID = /* ID of element that matching global element 10 */;
    fespace_serial.GetElementVDofs(global_cellID, dofs_serial);
    fespace_parallel.GetElementVDofs(local_cellID, dofs_parallel);
}
```

In serial, everything is a "global ID". In parallel, MFEM treats *most* objects locally. So passing "element 10" to the parallel space does not (necessarily) mean the same element as passing "element 10" to the serial space.

Hope that helps!

Cheers,
Tom

**Comment by martemyev** [comment]
Hi Tom,

Yes, indeed. So, that leads me to another question:
How can I check if some global element belongs to this process? It would be actually better to ask - is there a map that has a link between a global cell ID from a serial mesh to a process that owns it. That map would be super helpful in my code - because apparently I have some problems due to this reordering of DoF indices in the parallel finite element spaces, and using that map I could recover correct indices from the processes that own some global cells that I need.

Thanks,
Mikhail

**Comment by trb413** [comment]
Hi Mikhail,

So, this map does exist, but it is (usually) hidden from the user. By default, it is created in the call to the ParMesh constructor and not exposed to the user. You may, however, do that bit yourself. Modifying your earlier code:

```cxx
MPI_Init(&argc, &argv);

int myid, nproc;
MPI_Comm_rank(MPI_COMM_WORLD, &myid);
MPI_Comm_size(MPI_COMM_WORLD, &nproc);

Mesh mesh(/*parameters*/);
int * partitioning = mesh.GeneratePartitioning(nproc);
// Element ii gets sent to rank partitioning[ii]

ParMesh par_mesh(MPI_Comm, &mesh,partitioning);

/* etc */

delete[] partitioning;
```

This assumes you are not using a nonconforming mesh; in that case, partitioning is handled differently. But it will work for the conforming mesh path, and it will do exactly the same thing as if you just use the two-argument construction for your ParMesh.

Cheers,
Tom

**Comment by martemyev** [comment]
Hi Tom,

Thank you again for the help! I finally found a bug on my side and now I think everything works correctly.

Cheers,
Mikhail
